Userland/Libraries/LibWeb/HTML/Worker.h

```diff
@@ -0,0 +1,260 @@
+        //auto& window = verify_cast<Bindings::WindowObject>(realm->global_object()).impl();
+
+        // 3. Let settings object be a new environment settings object whose algorithms are defined as follows:
+        // NOTE: See the functions defined for this class.
+        auto settings_object = adopt_own(*new WorkerEnvironmentSettingsObject(document, execution_context));
+
+        // FIXME: 4. If reservedEnvironment is non-null, then:
+        // FIXME:    1. Set settings object's id to reservedEnvironment's id, target browsing context to reservedEnvironment's target browsing context, and active service worker to reservedEnvironment's active service worker.
+        // FIXME:    2. Set reservedEnvironment's id to the empty string.
+
+        // FIXME: 5. Otherwise, set settings object's id to a new unique opaque string, settings object's target browsing context to null, and settings object's active service worker to null.
+        settings_object->target_browsing_context = nullptr;
+
+        // FIXME: 6. Set settings object's creation URL to creationURL, settings object's top-level creation URL to topLevelCreationURL, and settings object's top-level origin to topLevelOrigin.
+        //settings_object->creation_url = creation_url;
+
+        // 7. Set realm's [[HostDefined]] field to settings object.
+        realm->set_host_defined(move(settings_object));
+
+        return static_cast<WorkerEnvironmentSettingsObject*>(realm->host_defined());
+    }
+    virtual ~WorkerEnvironmentSettingsObject() override = default;
+
+    RefPtr<DOM::Document> responsible_document() override { return m_document; }
+    String api_url_character_encoding() override { return m_document->encoding_or_default(); }
+    AK::URL api_base_url() override { return m_document->url(); }
+    Origin origin() override { return m_document->origin(); }
+    CanUseCrossOriginIsolatedAPIs cross_origin_isolated_capability() override { TODO(); }
+
+private:
+    NonnullRefPtr<DOM::Document> m_document;
+};
+
+class WorkerDebugConsoleClient final
+    : public JS::ConsoleClient
+    , public RefCounted<WorkerDebugConsoleClient>
+    , public Weakable<WorkerDebugConsoleClient> {
+public:
+    WorkerDebugConsoleClient(JS::Console& console)
+        : ConsoleClient(console)
+    {
+    }
+
+    virtual void clear() override
+    {
+        dbgln("\033[3J\033[H\033[2J");
+        m_group_stack_depth = 0;
+        fflush(stdout);
+    }
+
+    virtual void end_group() override
+    {
+        if (m_group_stack_depth > 0)
+            m_group_stack_depth--;
+    }
+
+    // 2.3. Printer(logLevel, args[, options]), https://console.spec.whatwg.org/#printer
+    virtual JS::ThrowCompletionOr<JS::Value> printer(JS::Console::LogLevel log_level, PrinterArguments arguments) override
+    {
+        String indent = String::repeated("  ", m_group_stack_depth);
+
+        if (log_level == JS::Console::LogLevel::Trace) {
```

Comment:
Unfortunate we have to copy all of this code, but not your problem.
However can we at least move this definition out of the header?